### PR TITLE
Added subset_metric, 4 tests and a test dataset specifically for test…

### PR DESCRIPTION
**Pull request by janvdvegt**
Not sure about automatically returning NaN for empty slices and throwing a warning. Decided to throw an exception in case the subset_method does not return a 1-dimensional array-like.

### CODE REVIEW

File: tests/test_metrics/test_subset_metric.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    model = RandomForestClassifier().fit(X, y)
+    with warnings.catch_warnings(record=True) as w:
+        # Cause all warnings to always be triggered.
+        warnings.simplefilter("always")
+        # Trigger a warning.
+        accuracy_x1_0(estimator=model, X=X, y_true=y)
+        assert issubclass(w[-1].category, RuntimeWarning)
+
+
+def test_wrong_subset_dimensions_raise_value_error(slicing_classification_dataset):
```
Comment (koaning) [inline tests/test_metrics/test_subset_metric.py]: if possible, could you add a test to confirm that the metric also works in a pipeline + gridsearch? the lambda functions cannot be pickled so you may need to declare an actual predicate function instead.

File: sklego/metrics.py [hunk 2]
```diff
@@ -152,3 +154,36 @@ def impl(estimator, X, y_true):
...
+    is a method that is passed the corresponding X and y_true and returns a one-dimensional
+    boolean vector where every element corresponds to a row in the data. Only the elements
+    with a True value are taken into account for the passed metric, representing a filter.
+
+    This allows users to have an easy approach to measuring metrics over different slices of
+    the population which can give insights into the model performance, either specifically for
+    fairness or in general.
+
+    Usage:
+    `subset_metric(lambda X, y_true: subset_method, metric_method)(clf, X, y)`
```
[the remark this answers is not in the document]
Comment (janvdvegt) [reply sklego/metrics.py]: Ok I will change the example. 

I included `y` so that you could use the method to also slice on outcomes. This way you could - for example - directly ask what the MSE is for predicted heart rates of women who fall in the high heart rate category.

File: tests/test_metrics/test_subset_metric.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+from sklearn.metrics import accuracy_score
+
+from sklego.metrics import subset_metric
+
+
+def test_subset_metric_accuracy_pandas(slicing_classification_dataset):
+    accuracy_x1_0 = subset_metric(lambda X, y_true: X['x1'] == 0, accuracy_score)
+    accuracy_x1_1 = subset_metric(lambda X, y_true: X['x1'] == 1, accuracy_score)
+    X, y = slicing_classification_dataset
+    model = RandomForestClassifier().fit(X, y)
```
[the remark this answers is not in the document]
Comment (janvdvegt) [reply tests/test_metrics/test_subset_metric.py]: Ok I will look into it, I didn't like my approach too much but my toy example should work always which I included it

File: tests/test_metrics/test_subset_metric.py (hunk 1 above)
Comment (janvdvegt) [reply tests/test_metrics/test_subset_metric.py]: Ok I will add it

File: sklego/metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (janvdvegt) [reply sklego/metrics.py]: I kept in the `y` for now, processed all the other feedback and updated the PR. If this choice of including `y` is fine then I think it is ready.

File: tests/test_metrics/test_subset_metric.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+from sklearn.metrics import accuracy_score
+from sklearn.pipeline import make_pipeline
+from sklearn.model_selection import GridSearchCV
+
+from sklego.preprocessing import ColumnSelector
+
+from sklego.metrics import subset_score
+
+
+class DisabledCV:
```
Comment (MBrouns) [inline tests/test_metrics/test_subset_metric.py]: Wouldn't setting the seed on an existing CV give the same result?

File: sklego/metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MBrouns) [reply sklego/metrics.py]: I think including `y` is a good idea. For example the `equal_opportunity_score` requires access to `y` and I could see more metrics having the same requirements.

File: sklego/metrics.py [hunk 5]
```diff
@@ -152,3 +154,37 @@ def impl(estimator, X, y_true):
...
+        mask = subset_picker(X, y_true)
+        if isinstance(mask, np.ndarray):
+            if len(mask.shape) > 1:
+                raise ValueError("`subset_picker` should return 1-dimensional numpy array or Pandas series")
+        if np.sum(mask) == 0:
+            warnings.warn(f"No samples in subset, returning NaN", RuntimeWarning)
+            return np.nan
+        X = X[mask]
+        y_pred = estimator.predict(X)
+        y_true = y_true[mask]
```
Comment (MBrouns) [inline sklego/metrics.py]: I think I'd having this `y_true` not shadow the `y_true` argument name. We could change it to:

```
return score(y_true[mask], y_pred[mask], **kwargs)
```

at a slight computational cost.

File: sklego/metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (koaning) [reply sklego/metrics.py]: fair enough.

**Comment by koaning** [comment]
Once the final comments from @MBrouns are in we could merge this.

**Comment by koaning** [comment]
This is a friendly reminder, @janvdvegt any progress on this :) ?

**Comment by koaning** [comment]
Hi @janvdvegt, this is another friendly reminder? Any work on this? We can also wrap this up if you prefer.

File: tests/test_metrics/test_subset_metric.py (hunk 4 above)
Comment (janvdvegt) [reply tests/test_metrics/test_subset_metric.py]: That is true but not guaranteed to stay exactly the same if they change the way they use the RNG. Plus it's harder to reason about because you have to see the actual output before understanding what it will. But I'm ok with changing this, what do you think?

**Comment by janvdvegt** [comment]
Hi all, sorry for the late response, I've been quite busy and I'm subscribed to some large repos on Github which made me miss these reminders. I had a bit of issues with updating my forked repository but I think it worked now, I resolved two of the comments and asked about the third one from @MBrouns , is the PR updated now that I pushed to my forked master?

**Comment by koaning** [comment]
I think this looks good now. In the near future odds are that I will use this feature as a part of the documentation before we do a new version release. Thanks! Also, you may now ask for a laptop-sticker when you're at the GDD office.
